File: c/src/mbug_2151.c

//==============================================================================
/* MBUG-2151 interface implementation

		Mostly identical to mbug_2165.c, but without carrier modulation.
		Any changes in this file should also be applied in mbug_2165.c.
*/
//==============================================================================

#include <stdlib.h>
#include <string.h>
#include <ctype.h>  // toupper()
#include "mbug_2151.h"

//------------------------------------------------------------------------------
// Parameter numbers (keep private, may change in future revisions)

enum  mbug_2151_param {
	PARAM_START = 0x10,
	PARAM_RESET = 0xFF,
	PARAM_ITER  = 0x01,
	PARAM_CLOCK_DIV = 0x02,
	PARAM_BASE_CLOCK = 0x05,
	PARAM_BUSY = 0x70,
	PARAM_SEQ_LEN = 0x71
};

//------------------------------------------------------------------------------
// Get a list of all available mbug-2151 devices.
const mbug_device_list  mbug_2151_list( void )
{
	return mbug_get_device_list(2151);
}

//------------------------------------------------------------------------------
// Device base clock.  This is assumed to be identical for
//   all mbug-2151 devices and only read the very first time a device is opened.
//   Currently, _base_clock = _mod_clock = 4 MHz.

static int base_clock = 0;

//------------------------------------------------------------------------------
// Open a device specified by it's serial number (as int,
// last digits of the serial number are matched only).
mbug_device mbug_2151_open( int serial_num )
{
	mbug_device dev = mbug_open_int(2151, serial_num);
	if ( (dev != NULL) && (base_clock == 0) )
	{
		base_clock = mbug_2151_get_base_clock( dev );
	}
	return dev;
}

//------------------------------------------------------------------------------
// Open a device specified by it's id string, as returned
// by mbug_2151_list()
mbug_device mbug_2151_open_str( const char *id )
{
	mbug_device dev;
	unsigned int type = mbug_type_from_id( id );
	if (type != 2151) return NULL;
	dev = mbug_open_str( id );
	if ( (dev != NULL) && (base_clock == 0) )
	{
		base_clock = mbug_2151_get_base_clock( dev );
	}
	return dev;
}

//------------------------------------------------------------------------------
// Close a previously opened device.
void mbug_2151_close( mbug_device dev )
{
	mbug_close(dev);
}

//------------------------------------------------------------------------------
// Private helper function:
//   Generic I/O command: Send cmd, receive integer parameter
//   Return values <0 indicate errors.
static int io( mbug_device dev,  unsigned char* out,  size_t size )
{
	unsigned char in[8] = {0};
	short status = 0;
	int value = 0;

	if (mbug_write( dev, out, size ) <0)
		return -1;
	if (mbug_read( dev, in, sizeof(in) ) <0)
		return -1;

	// First 2 bytes contain the return status (signed 16 bit)
	status = *(short*) in;
	// Next 4 bytes contain the returned output value
	value = *(int*) (in+2);
	if (status<0)  return status;
	else  return value;
}

//------------------------------------------------------------------------------
// Private helper function:
//   Set parameter value.
static int set_param( mbug_device dev, enum mbug_2151_param param, unsigned int value )
{
	unsigned char out[8] = {0xB0, param, 0, 0, 0, 0, 0, 0};
	memcpy( out+2, (void*)&value, 4 );
	return io( dev, out, 6);
}

//------------------------------------------------------------------------------
// Private helper function:
//   Get parameter value.
static int get_param( mbug_device dev, enum mbug_2151_param param )
{
	unsigned char out[2] = {0xC0, param};
	return io( dev, out, 2);
}

//------------------------------------------------------------------------------
// Set transmission sequence (max 255 bytes, selectable transmission mode)
int mbug_2151_set_sequence( mbug_device dev, unsigned char* data, size_t nbytes, mbug_2151_tx_mode tx_mode )
{
	unsigned char out[64] = {0xB0, 0};
	unsigned char* pd = data;
	size_t  s, smax;

	if (nbytes>255)  nbytes=255;
	if (tx_mode > TX_MODE_TIMED_16)
		tx_mode = TX_MODE_BITSTREAM;

	smax = 64 - 4;  // Maximum data length per packet

	// First packet: Set sequence
	// Command format: {0xB0, 0, mode<<4, len, data[:] }
	s = nbytes>smax ? smax : nbytes;
	out[2] = tx_mode<<4;  out[3] = s;
	memcpy( out+4, pd, s );
	nbytes -= s;  pd += s;
	if (io( dev, out, s+4 ) <0)
		return -1;

	// Subsequent packages: Append sequence
	// Command format: {0xB0, 0, 0x01, len, data[] }
	while (nbytes>0)
	{
		s = nbytes>smax ? smax : nbytes;
		out[2] = 0x01;  out[3] = s;
		memcpy( out+4, pd, s );
		nbytes -= s;  pd += s;
		if (io( dev, out, s+4 ) <0)
			return -1;
	}
	return 0;
}

//------------------------------------------------------------------------------
// Set transmission sequence in bitstream mode (max 255 bytes)
int mbug_2151_set_seq_bitstream( mbug_device dev, unsigned char* data, size_t nbytes )
{
	return mbug_2151_set_sequence( dev, data, nbytes, TX_MODE_BITSTREAM );
}

//------------------------------------------------------------------------------
// Set transmission sequence in 8-bit time mode (max 255 bytes)
int mbug_2151_set_seq_times_8bit( mbug_device dev, unsigned char* data, size_t nbytes )
{
	return mbug_2151_set_sequence( dev, data, nbytes, TX_MODE_TIMED_8 );
}

//------------------------------------------------------------------------------
// Set transmission sequence in 16-bit time mode (max 127 items)
int mbug_2151_set_seq_times_16bit( mbug_device dev, unsigned short* data, size_t nitems )
{
	return mbug_2151_set_sequence( dev, (unsigned char*)data, nitems*2, TX_MODE_TIMED_16 );
}

//------------------------------------------------------------------------------
// Get the size (in bytes) of the currently set sequence
int mbug_2151_get_seq_length( mbug_device dev )
{
	return get_param( dev, PARAM_SEQ_LEN );
}

//------------------------------------------------------------------------------
// Set number of transmission sequence iterations,
// 0: infinite, Maximum: 2**16-1
int mbug_2151_set_iterations( mbug_device dev, unsigned int niter )
{
	if (niter >= 1<<16 )  niter = (1<<16)-1;
	return set_param( dev, PARAM_ITER, niter );
}

//------------------------------------------------------------------------------
// Get the currently set number of transmission sequence iterations,
// 0: infinite
int mbug_2151_get_iterations( mbug_device dev )
{
	return get_param( dev, PARAM_ITER );
}

//------------------------------------------------------------------------------
// Get the devices base transmission clock frequency in Hz
int mbug_2151_get_base_clock( mbug_device dev )
{
	return get_param( dev, PARAM_BASE_CLOCK );
}

//------------------------------------------------------------------------------
// Set clock divider. => Bitrate = base_clock/div
// Minimum div is 10 for bitstream mode, 1 for timed modes,
// Maximum div is 2**16-1 (65535).
int mbug_2151_set_clock_div( mbug_device dev, unsigned int div )
{
	if (div == 0)  div = 1;
	if (div >= 1<<16 )  div = (1<<16)-1;
	return set_param( dev, PARAM_CLOCK_DIV, div );
}

//------------------------------------------------------------------------------
// Get the currently set transmission clock divider.
int mbug_2151_get_clock_div( mbug_device dev )
{
	return get_param( dev, PARAM_CLOCK_DIV );
}

//------------------------------------------------------------------------------
// Set the transmission rate (via clock div).
// Minimum rate is 61sec.
// Maximum rate is 400kHz for bitstream mode, 4Mhz for timed modes.
// Returns the actually set bitrate or <0 in case of an error.
double mbug_2151_set_bitrate( mbug_device dev, double freq )
{
	int div;
	div = (int)(0.5 + 1.0*base_clock/freq);
	if (div >= 1<<16)  div = (1<<16)-1;
	if( mbug_2151_set_clock_div( dev, div ) < 0)
		return -1;
	return (int)(0.5 + 1.0*base_clock/div);
}

//------------------------------------------------------------------------------
// Set the transmission time base (in seconds, via clock div).
// Minimum timebase is 2.5us for bitstream mode, 0.25us for timed modes.
// Maximum timebase is 16.38375ms.
// Resolution is 0.25us.
// Returns the actually set timebase or <0 in case of an error.
double mbug_2151_set_timebase( mbug_device dev, double interval )
{
	int div;
	div = (int)(0.5 + 1.0*interval*base_clock);
	if (div >= 1<<16)  div = (1<<16)-1;
	if( mbug_2151_set_clock_div( dev, div ) < 0)
		return -1;
	return (int)(0.5 + 1.0*div/base_clock);
}

//------------------------------------------------------------------------------
// Start transmission.
int mbug_2151_start( mbug_device dev )
{
	return set_param( dev, PARAM_START, 0x0001 );
}


//------------------------------------------------------------------------------
// Stop transmission
//	 force = 0: Stop gracefully after ongoing iteration
//	 force = 1: Stop instantly
int mbug_2151_stop( mbug_device dev, int force )
{
	return set_param( dev, PARAM_START, force ? 0x0100 : 0x0000  );
}

//------------------------------------------------------------------------------
// Stop transmission gracefully, i.e. after the current iteration.
int mbug_2151_stop_gracefully( mbug_device dev )
{
	return set_param( dev, PARAM_START, 0x0000 );
}

//------------------------------------------------------------------------------
// Stop transmission instantly.
int mbug_2151_stop_immediately( mbug_device dev )
{
	return set_param( dev, PARAM_START, 0x0100 );
}

//------------------------------------------------------------------------------
// Reset transmission and all parameters
int mbug_2151_reset( mbug_device dev )
{
	return set_param( dev, PARAM_RESET, 0 );
}

//------------------------------------------------------------------------------
// Get busy state. >0 if transmission is in progress
int mbug_2151_get_busy( mbug_device dev )
{
	return get_param( dev, PARAM_BUSY );
}

//------------------------------------------------------------------------------

int mbug_2151_ab440s_str_to_addr( const char *str ) {
	int addr = 0;
	int i = 0;

	for(i = 0; str[i]; i++) {
		if(str[i] >= '1' && str[i] <= '5') {
			addr |= 1<<(9-str[i]+'1');
		} else if(toupper(str[i]) >= 'A' && toupper(str[i]) <= 'E') {
			addr |= 1<<(4-toupper(str[i])+'A');
		}
	}

	return addr;
}

int mbug_2151_ab440s_switch_addr( mbug_device dev, int addr, int state ) {
	unsigned char seq[16];
	int i;

	if(mbug_2151_set_bitrate(dev, 3125) == -1) {
		return -1;
	}

	if(mbug_2151_set_iterations(dev, 3) == -1) {
		return -1;
	}

	for(i = 0; i < 12; i++) {
		seq[i] = 0x71;
	}

	if(state == 1) {
		seq[10] = 0x11;
	} else {
		seq[11] = 0x11;
	}
	seq[12] = 1;
	seq[13] = seq[14] = seq[15] = 0;
	for(i = 0; i < 10; i++) {
		seq[9-i] = (addr & (1<<i)) ? 0x11 : 0x71;
	}
	if(mbug_2151_set_sequence(dev, seq, 16, TX_MODE_BITSTREAM) == -1) {
		return -1;
	}
	return mbug_2151_start(dev);
}

int mbug_2151_ab440s_switch_str( mbug_device dev, const char *addr, int state ) {
	return mbug_2151_ab440s_switch_addr(dev, 
										mbug_2151_ab440s_str_to_addr(addr),
										state);
}

int mbug_2151_dmv7008_cmd_addr(mbug_device dev, int syscode, int channel, int cmd) {
	int i;
	char seq[59];
	int bits;
	const unsigned char cmdbits[5][4] = {
		{ 0xE1, 0x70, 0xEB, 0x7A },
		{ 0x00, 0x11, 0x0A, 0x1B },
		{ 0x82, 0x93, 0x88, 0x99 },
		{ 0x41, 0x50, 0x4B, 0x5A },
		{ 0xC3, 0xD2, 0xC9, 0xD8 }
	};

	syscode = syscode & 0xFFF;
	if(channel < 0 || channel > 4) {
		return -1;
	}

	bits = (1 << 20) | (syscode << 8) | (cmdbits[channel][cmd] << 0);

	for(i = 0; i < 21; i++) {
		seq[i] = ((bits >> (20-i)) & 1) ? 0xC0 : 0xFC;
	}
	for( ; i < 59; i++) {
		seq[i] = 0;
	}

	if(mbug_2151_set_bitrate(dev, 4167) == -1) {
		return -1;
	}

	if(mbug_2151_set_iterations(dev, 4) == -1) {
		return -1;
	}

	if(mbug_2151_set_sequence(dev, seq, 59, TX_MODE_BITSTREAM) == -1) {
		return -1;
	}

	return mbug_2151_start(dev);
}

int mbug_2151_dmv7008_cmd_str(mbug_device dev, const char *addr, int cmd) {
	static int syscode = 0;
	int channel;

	char *colon = strchr(addr, ':');
	if(colon != NULL) {
		syscode = atoi(addr);
		channel = atoi(colon+1);
	} else {
		channel = atoi(addr);
	}

	return mbug_2151_dmv7008_cmd_addr(dev, syscode, channel, cmd);
}

int mbug_2151_rs200_cmd_addr(mbug_device dev, int syscode, int channel, int cmd) {
	int i;
	unsigned short seq[26*2];
	unsigned int data;

	if(syscode < 0 || syscode > 255) {
		return -1;
	}

	if(channel < 0 || channel == 4 || channel > 5) {
		return -1;
	}

	data = (channel << 4) | (cmd << 7) | ((1^cmd) << 2);
	if(channel == 1 || channel == 2) {
		data ^= 1<<7;
	}
	data = (0x19 << 20)
		| (syscode << 12)
		| (data << 4)
		| (0x9 + syscode + (syscode>>4) + data + (data>>4)) & 0x0F;

	for(i = 0; i < 26; i++) {
		if(data & (1<<(25-i))) {
			seq[2*i+0] = 600;
			seq[2*i+1] = 3400;
		} else {
			seq[2*i+0] = 1300;
			seq[2*i+1] = 3400;
		}
	}
	seq[26*2-1] += 30000;
	
	if(mbug_2151_set_timebase(dev, 1e-6) == -1) {
		return -1;
	}

	if(mbug_2151_set_iterations(dev, 4) == -1) {
		return -1;
	}

	if(mbug_2151_set_seq_times_16bit(dev, seq, 26*2) == -1) {
		return -1;
	}

	return mbug_2151_start(dev);
}

int mbug_2151_rs200_cmd_str(mbug_device dev, char *addr, int cmd) {
	static int syscode = 0;
	int channel;

	char *colon = strchr(addr, ':');
	char *chanptr;
	if(colon != NULL) {
		syscode = atoi(addr);
		chanptr = colon+1;
	} else {
		chanptr = addr;
	}

	if(toupper(*chanptr) == 'A'
	   || toupper(*chanptr) == 'M') {
		channel = 5;
	} else {
		channel = atoi(chanptr) - 1;
	}

	return mbug_2151_rs200_cmd_addr(dev, syscode, channel, cmd);
}

File: py/mbug_dev/mbug_2151.py

from mbug_base import *
import mbug_2151_target
import time as _time

#==========================================================================
class mbug_2151(mbug):
    """ Represents an MBUG-2151 433MHz transmitter"""
    
    _type = 2151

    def __init__(self, serial=None):
        mbug.__init__(self, self._type, serial)
        self._base_clock = self.get_base_clock()
        self.target._dev = self

    def _iocmd(self, cmd):
        self._write( cmd )
        ret = self._read()
        status = (ret[0] + ret[1]*256) - 2**16*(ret[1]&0x80>0)   # signed 16bit
        if status!=0:
            errmsg = {-1: 'Unknown command',
                }.get( status, 'Undefined error' )
            raise RuntimeError('mbug_2165: Communication Error: %s (%d)' % (errmsg,status) )
        return ret[2:]

    def set_sequence_bitstream(self, data):
        """ Set a transmission byte sequence in bitstream mode. The data will be 
        transmitted bit per bit, starting with the least significant bit of the 
        first byte. The maximum length is 255 bytes at the moment. """
        nbytes = len(data)
        if nbytes>255: raise ValueError('Maximum sequence length is 255.')
        s = self._size_out-4 # max data size per packet
        # First packet: set_sequence
        d, data = data[:s], data[s:]
        self._iocmd( bytearray(( 0xB0, 0x00, 0x00, len(d) )) + bytearray(d) )
        #Subsequent packets: append_sequence
        while (len(data)>0):
            d, data = data[:s], data[s:]
            self._iocmd( bytearray(( 0xB0, 0x00, 0x01, len(d) )) + bytearray(d) )
        return
        
    def set_sequence_times(self, data, unit='ms', clock_div=0, bits=0):
    	""" Set a transmission sequence in timed mode. The sequence is interpreted 
    	as consecutive on/off intevalls (starting with on). The default unit is
    	milliseconds (can be changed with the unit parameter). Suitable clock
    	dividers and word length (8 or 16 bit) are chose automatically from the 
    	data if not passed explicitely. If set manually, the on/off time intervals
    	equate as t*clkdiv/base_clock. Minimum interval = ??? XXXX"""
    	print ":TODO: NOT IMPLEMENTED YET"

    def set_sequence_times_8(self, data):
        nbytes = len(data)
        if nbytes>255: raise ValueError('Maximum sequence length is 255.')
        s = self._size_out-4 # max data size per packet
        # First packet: set_sequence
        d, data = data[:s], data[s:]
        self._iocmd( bytearray(( 0xB0, 0x00, 0x10, len(d) )) + bytearray(d) )
        #Subsequent packets: append_sequence
        while (len(data)>0):
            d, data = data[:s], data[s:]
            self._iocmd( bytearray(( 0xB0, 0x00, 0x01, len(d) )) + bytearray(d) )
        return

    def set_sequence_times_16(self, data):
        nitems = len(data)
        if nitems>127: raise ValueError('Maximum sequence length is 127.')
        bytes = []
        for item in data: bytes += [ item&0xFF, (item>>8)&0xFF]
        s = self._size_out-4 # max data size per packet
        # First packet: set_sequence
        d, bytes = bytes[:s], bytes[s:]
        self._iocmd( bytearray(( 0xB0, 0x00, 0x20, len(d) )) + bytearray(d) )
        #Subsequent packets: append_sequence
        while (len(bytes)>0):
            d, bytes = bytes[:s], bytes[s:]
            self._iocmd( bytearray(( 0xB0, 0x00, 0x01, len(d) )) + bytearray(d) )
        return

    def set_sequence(self, data, seq_type):
        """Set a transmission sequence of the type specified by seq_type
        ('bitstream', 'timed_8', 'timed_16')."""
        if seq_type=='bitstream':
            return self.set_sequence_bitstream(data)
        elif seq_type.startswith('timed_8'):
            return self.set_sequence_times_8(data)
        elif seq_type.startswith('timed_16'):
            return self.set_sequence_times_16(data)
        else: raise ValueError('Invalid sequence type')
    
    def set_iterations(self, niter):
        """ Set the number of transmission iterations.
        0: infinite, Maximum: 2**16-1 at the moment. """
        if niter=='inf': niter=0
        if niter<0 or niter>=1<<16:  raise ValueError('Invalid number of iterations.')
        self._iocmd( (0xB0, 0x01, niter&0xFF, (niter>>8)&0xFF, 0x00, 0x00 ) )
        return

    def set_clock_div(self, div):
        """ Set the clock divider. Bitrate = base_clock/div, Time = base_time*div
        Minimum div in bitstream mode is 10, maximum div is 2**16-1. """
        div = int(div)
        if div<=0 or div>=1<<16:  raise ValueError('Invalid number of iterations.')
        self._iocmd( (0xB0, 0x02, div&0xFF, (div>>8)&0xFF, 0x00, 0x00 ) )
        return

    def set_bitrate(self, freq):
        """ Set the transmission rate (in Hz, via clock div). Minimum rate is 61sec.
        Maximum rate is 400kHz for bitstream mode, 4Mhz for timed modes.
        Returns the actually set bitrate."""
        if freq<=0: raise ValueError('Invalid bitrate.')
        div = round(1.*self._base_clock/freq)
        if div>2**16-1: div=2**16-1;
        rfreq = self._base_clock/div
        if abs(1.*rfreq/freq-1.)>0.01:  # Deviation > 1%
            print "Warning. Set bitrate to %.1e."%(rfreq)
        self.set_clock_div(div)
        return rfreq

    def set_timebase(self, interval):
        """ Set the transmission base interval (in seconds, via clock div).
        Minimum interval is 2.5us for bitstream mode, 0.25us for timed modes.
        Maximum base interval is 16.38375ms. Resolution is 0.25us.
        Returns the actually set interval."""
        div = round(1.*self._base_clock*interval)
        if div>2**16-1: div=2**16-1;
        self.set_clock_div(div)
        return int(round(1.0*div/self._base_clock))

    def start(self):
        """ Start transmission. """
        self._iocmd( (0xB0, 0x10, 0x01) )
        return

    def stop(self, instantly=0):
        """ Stop transmission.
        instantly==0: Stop gracefully after the current iteration.
        instantly==1: Stop instantly. (Discards remaining sequence bytes.)"""
        if instantly: self._iocmd( (0xB0, 0x10, 0x00, 0x01) )
        else:         self._iocmd( (0xB0, 0x10, 0x00, 0x00) )
        return

    def reset(self):
        """ Reset transmission and all parameters. """
        self._iocmd( (0xB0, 0xFF) )
        return

    def get_iterations(self):
        """ Query the number of transmission iterations (0:infinite)."""
        r = self._iocmd( (0xC0, 0x01) )   # Query base clock
        return sum( r[i]<<(i*8) for i in range(4) )

    def get_clock_div(self):
        """ Get the bit clock divider. Bitrate = base_clock/div."""
        r = self._iocmd( (0xC0, 0x02) )   # Query base clock
        return sum( r[i]<<(i*8) for i in range(4) )

    def get_mod_clock_div(self):
        """ Get the modulation clock divider. mod_freq = mod_base_clock/mod_div."""
        r = self._iocmd( (0xC0, 0x03) )   # Query base clock
        return sum( r[i]<<(i*8) for i in range(4) )
    
    def get_base_clock(self):
        """ Query the devices base bit clock frequency in Hz. """
        r = self._iocmd( (0xC0, 0x05) )   # Query base clock
        return sum( r[i]<<(i*8) for i in range(4) )

    def get_mod_base_clock(self):
        """ Query the devices modulation base clock frequency in Hz. """
        r = self._iocmd( (0xC0, 0x06) )   # Query base clock
        return sum( r[i]<<(i*8) for i in range(4) )
        
    def get_busy(self):
        """ Get the busy state. True if transmission is in progress. """      
        r = self._iocmd( (0xC0, 0x70) )   # Query tx_busy
        return bool( r[0] )

    def get_sequence_length(self):
        """ Get the size of the stored sequence in bytes """
        r = self._iocmd( (0xC0, 0x71) )  # Query seq_length
        return int( r[0]+r[1]*256 )
    
#----------------------------------------------------------------------------

    class target:
        """ A collection of target devices addressable by the mbug_2151.
        The target device must be created from an mbug_2151 instance, which
        will be used as controller."""
        
        _dev = None     # The controller device
        
        # Target device factories
        def AB440S(self, addr=None):
            return mbug_2151_target.AB440S(dev=self._dev, addr=addr)
        def HE302EU(self, addr=None):
            return mbug_2151_target.HE302EU(dev=self._dev, addr=addr)
        def DMV7008(self, addr=None):
            return mbug_2151_target.DMV7008(dev=self._dev, addr=addr)
        def GT7008(self, addr=None):
            return mbug_2151_target.GT7008(dev=self._dev, addr=addr)
        def FIF4280(self, addr=None):
            return mbug_2151_target.FIF4280(dev=self._dev, addr=addr)
        def IKT201(self, addr=None):
            return mbug_2151_target.IKT201(dev=self._dev, addr=addr)
        def RS200(self, addr=None):
            return mbug_2151_target.RS200(dev=self._dev, addr=addr)
        
    target = target()   # Make this an instance 


#==========================================================================

def mbug_2151_test():
    # Print device list
    devs = list_devices(2151)
    print "Attached MBUG-2151 devices:"
    print devs if devs!=[] else 'None'

    # Open devices
    for ser in devs:
        try:
            print "Device", ser
            dev = mbug_2151(ser)
            sw = dev.target.AB440S()
            sw['A'] = 1
            _time.sleep(1)
            sw['A'] = 0
        finally:
            try: dev.close()
            except: pass

#---------------------------------------------------------------------------
if __name__ == "__main__":
    pass
    mbug_2151_test()
    
#---------------------------------------------------------------------------


